From 7d2135e75ca81902d6d22cee568ab9430de05441 Mon Sep 17 00:00:00 2001
From: Subrata Chatterjee <schatterjee@habana.ai>
Date: Wed, 27 Jul 2022 11:54:18 +0300
Subject: [PATCH] [SW-95911] gaudi2: bist: sram pll config

sram bist should run on functional clock than
on reference clock. So added code to configure SRAM
clocks into functional frequency (350Mhz).

Change-Id: Ibe45b8ac9b2994c58d6991ac0f0a6dc8f6b2ad72
---

diff --git a/arch/gaudi2/arc/asic.c b/arch/gaudi2/arc/asic.c
index 2430a35..0700b1e 100644
--- a/arch/gaudi2/arc/asic.c
+++ b/arch/gaudi2/arc/asic.c
@@ -94,6 +94,30 @@
 	}
 };
 
+static struct hl_pll_cfg gaudi2_sram_pll_cfg[] = {
+	{
+		.name = "DCORE2_SRAM",
+		.base = mmDCORE2_XBAR_BANK_PLL_CTRL_BASE,
+		.nr = 1, .nf = 63, .od = 3, .nb = 3,	/* 350 */
+		.div_factor = { 0, -1, -1, -1 },
+		.div_sel = { HL_PLL_PLL_CLK,
+			     HL_PLL_REF_CLK,
+			     HL_PLL_REF_CLK,
+			     HL_PLL_REF_CLK
+		},
+	}, {
+		.name = "DCORE3_SRAM",
+		.base = mmDCORE3_XBAR_BANK_PLL_CTRL_BASE,
+		.nr = 1, .nf = 63, .od = 3, .nb = 3,	/* 350 */
+		.div_factor = { 0, -1, -1, -1 },
+		.div_sel = { HL_PLL_PLL_CLK,
+			     HL_PLL_REF_CLK,
+			     HL_PLL_REF_CLK,
+			     HL_PLL_REF_CLK
+		},
+	}
+};
+
 void hl_arch_pll_init(void)
 {
 	/* In case of secureboot flow psoc pll needs to be configured */
@@ -143,6 +167,11 @@
 
 	for (i = 0; i < (size / 4); i++)
 		*(dst + i) = *(src + i);
+}
+
+static void gaudi2_sram_pll_init(void)
+{
+	hl_pll_init(gaudi2_sram_pll_cfg, ARRAY_SIZE(gaudi2_sram_pll_cfg));
 }
 
 static int gaudi2_sms_check_pattern(uint32_t value, uint32_t ones_mask, uint32_t zeros_mask)
@@ -324,6 +353,9 @@
 {
 	init_sram_bist_data();
 
+	/* Configure SRAM PLLs before running BIST operations */
+	gaudi2_sram_pll_init();
+
 	gaudi2_sram_self_repair(SRAM_SELF_REPAIR_OP_BIRA);
 	gaudi2_sram_self_repair(SRAM_SELF_REPAIR_OP_BISR);
 	gaudi2_sram_self_repair(SRAM_SELF_REPAIR_OP_BIST);
